src/native/simplekdf.rs

```diff
@@ -0,0 +1,280 @@
+            }
+            CKM_CONCATENATE_BASE_AND_DATA
+            | CKM_CONCATENATE_DATA_AND_BASE
+            | CKM_XOR_BASE_AND_DATA => {
+                if base_key_info.sensitive {
+                    tmpl.add_bool(CKA_SENSITIVE, &CK_TRUE);
+                }
+                if !base_key_info.extractable {
+                    tmpl.add_bool(CKA_EXTRACTABLE, &CK_FALSE);
+                }
+                if base_key_info.always_sensitive {
+                    tmpl.add_bool(CKA_ALWAYS_SENSITIVE, &CK_TRUE);
+                }
+                if base_key_info.never_extractable {
+                    tmpl.add_bool(CKA_NEVER_EXTRACTABLE, &CK_TRUE);
+                }
+                match &self.data {
+                    Some(d) => d.len(),
+                    None => return Err(CKR_MECHANISM_PARAM_INVALID)?,
+                }
+            }
+            _ => return Err(CKR_MECHANISM_INVALID)?,
+        };
+
+        let outlen = if self.mech == CKM_XOR_BASE_AND_DATA {
+            std::cmp::min(base_key_info.value.len(), other_data_len)
+        } else {
+            base_key_info.value.len() + other_data_len
+        };
+
+        let factory =
+            objfactories.get_obj_factory_from_key_template(tmpl.as_slice())?;
+
+        // check the length is compatible with the key type. Add default if missing
+        let keylen = match tmpl.find_attr(CKA_VALUE_LEN) {
+            Some(a) => usize::try_from(a.to_ulong()?)?,
+            None => {
+                match factory
+                    .as_secret_key_factory()?
+                    .recommend_key_size(outlen)
+                {
+                    Ok(len) => len,
+                    Err(_) => return Err(CKR_TEMPLATE_INCONSISTENT)?,
+                }
+            }
+        };
+
+        if keylen > outlen {
+            return Err(CKR_TEMPLATE_INCONSISTENT)?;
+        }
+
+        let mut secret = Vec::with_capacity(outlen);
+        match self.mech {
+            CKM_CONCATENATE_BASE_AND_KEY => {
+                secret.extend_from_slice(&base_key_info.value);
+                if let Some(k) = &self.key {
+                    secret.extend_from_slice(&k.value);
+                }
+                secret.resize(keylen, 0);
+            }
+            CKM_CONCATENATE_BASE_AND_DATA => {
+                secret.extend_from_slice(&base_key_info.value);
+                if let Some(d) = &self.data {
+                    secret.extend_from_slice(&d);
+                }
+                secret.resize(keylen, 0);
+            }
+            CKM_CONCATENATE_DATA_AND_BASE => {
+                if let Some(d) = &self.data {
+                    secret.extend_from_slice(&d);
+                }
+                secret.extend_from_slice(&base_key_info.value);
+                secret.resize(keylen, 0);
+            }
+            CKM_XOR_BASE_AND_DATA => {
+                if let Some(d) = &self.data {
+                    secret.extend_from_slice(&d);
```

Comment:
If this is null, then the result will be the empty string, shouldn't it be an error to have empty data ?